app/naive_bayes.py

```diff
@@ -0,0 +1,286 @@
+
+        :param links: list of links that agent could execute
+        """
+        dict_link_index_to_prob_success = dict()
+        dict_link_index_to_num_observations = dict()
+        links_insufficient_data = []
+
+        for index in range(len(links)):
+            cur_link = links[index]
+            # NOTE: link_query_features allows customizing which of 16 features to query by
+            # current selection of features with broad scope:
+            link_query_features = {
+                "Ability_ID": str(cur_link.ability.ability_id),
+                "Link_Facts": self._get_useful_facts(cur_link),
+                "Executor_Platform": str(cur_link.executor.platform)
+            }
+            prob_link_success, num_link_observations = self._get_link_success_probability(link_query_features)
+            if prob_link_success == None or str(cur_link.ability.ability_id) == "300157e5-f4ad-4569-b533-9d1fa0e74d74":
+                # Manual handle that if ability is compress staged directory then store as 
+                # not enough data and default it to atomic (push to end of execution) for better operation ordering
+                links_insufficient_data.append(cur_link)
+            else:
+                dict_link_index_to_prob_success[index] = prob_link_success
+                dict_link_index_to_num_observations[index] = num_link_observations
+        
+        if len(dict_link_index_to_prob_success.keys()) > 0:
+            indexBestLink = max(dict_link_index_to_prob_success, key=dict_link_index_to_prob_success.get)
+            probSuccessBestLink = dict_link_index_to_prob_success[indexBestLink]
+            if probSuccessBestLink >= self.min_prob_link_success:
+                if self.decision_logging:
+                    self.planning_svc.log.debug("Best link selected with probability of success " + "{:.2f}".format(probSuccessBestLink*100)
+                    + "%" + " based on " + str(dict_link_index_to_num_observations[indexBestLink]) +" observations")
+                return links[indexBestLink]
+            else:
+                if  len(links_insufficient_data) == 0:
+                    if self.decision_logging:
+                        self.planning_svc.log.debug("Skipping " + str(len(dict_link_index_to_prob_success.keys()))
+                        + " link(s) with insufficient likelihood of success")
+                    return None
+
+        # backup atomic ordering 
+        abil_id_to_link = dict()
+        for link in links_insufficient_data:
+            abil_id_to_link[link.ability.ability_id] = link
+        candidate_ids = set(abil_id_to_link.keys())
+        for ab_id in self.operation.adversary.atomic_ordering:
+            if ab_id in candidate_ids:
+                if self.decision_logging:
+                    self.planning_svc.log.debug("Link selected with backup atomic ordering")
+                return abil_id_to_link[ab_id]
```

Comment:
I think there may be too many returns in this method. I think the general rule we try to stick to is no more than two to a method. This may be a good opportunity to break out some of the code in these two blocks into their own methods that both return a link, and then `_get_best_link` can just do one return based on which method was executed.